Source/WebCore/loader/MixedContentChecker.cpp

```diff
@@ -76,25 +77,33 @@ static bool foundMixedContentInFrameTree(const LocalFrame& frame, const URL& url
     return false;
 }
 
-
-static void logWarning(const LocalFrame& frame, bool allowed, ASCIILiteral action, const URL& target)
+static void logConsoleWarning(const LocalFrame& frame, bool allowed, ASCIILiteral action, const URL& target)
 {
     const char* errorString = allowed ? " was allowed to " : " was not allowed to ";
     auto message = makeString((allowed ? "" : "[blocked] "), "The page at ", frame.document()->url().stringCenterEllipsizedToLength(), errorString, action, " insecure content from ", target.stringCenterEllipsizedToLength(), ".\n");
     frame.document()->addConsoleMessage(MessageSource::Security, MessageLevel::Warning, message);
 }
 
-bool MixedContentChecker::frameAndAncestorsCanDisplayInsecureContent(LocalFrame& frame, ContentType type, const URL& url)
+static void logConsoleWarningForUpgrade(const LocalFrame& frame, bool blocked, const URL& target)
+{
+    const char* errorString = !blocked ? "automatically upgraded and should" : "blocked and must";
+    auto message = makeString((!blocked ? "" : "[blocked] "), "The page at ", frame.document()->url().stringCenterEllipsizedToLength(), " requested insecure content from ", target.stringCenterEllipsizedToLength(), ". This content was ", errorString, " be served over HTTPS.\n");
+    frame.document()->addConsoleMessage(MessageSource::Security, MessageLevel::Warning, message);
+}
+
+static bool frameAndAncestorsCanDisplayInsecureContent(LocalFrame& frame, MixedContentChecker::ContentType type, const URL& url)
 {
+    if (frame.document()->settings().upgradeMixedContentEnabled())
+        return true;
     if (!foundMixedContentInFrameTree(frame, url))
         return true;
 
     RefPtr document = frame.document();
```

Comment:
Let's move this line before line 96.
Also, foundMixedContentInFrameTree was making sure document would not be null.
But this is no longer the case.
What is guaranteeing that frame.document() is not null?
Ditto below

Reply:
Good question, I see a few unchecked uses of frame.document in MixedContent.cpp but I don't see any reason why it is guaranteed to be non-null.

I put the check for `upgradeMixedContentEnabled` so we avoid walking the frame tree if the setting is enabled. I'd rather just add null checks.